City-of-Helsinki/hel-fi-drupal-grants
# fix: AU-1745: AtvSchema refactor: typedDataToDocumentContentWithWebform and getPropertySchema

# [AU-1745](https://helsinkisolutionoffice.atlassian.net/browse/AU-1745), [AU-1741](https://helsinkisolutionoffice.atlassian.net/browse/AU-1741)

## What was done
This pull request refactors the `typedDataToDocumentContentWithWebform` and `getPropertySchema` methods inside the `AtvSchema` class. Refactoring is the process of restructuring code, while not changing its original functionality. Hence, bugs that were found in both of these methods have been left untouched. This is also partially the reason why further refactoring of especially the `typedDataToDocumentContentWithWebform` is problematic.

### The getPropertySchema method
The `getPropertySchema` method has been refactored and extracted from the `AtvSchema` class into its own `PropertySchema` class. Previously, SonarCloud complained about the following issues regarding the method:

- The method had a cognitive complexity of 115.
- The method had too many return statements.
- The method had if statements that could be merged.
- The method had lines that were too long.

The new `PropertySchema` class and its methods fix all the mentioned issues. The new methods that have been implemented in the class are the following:

- `getPropertySchema()` (the new "main method" of the class)
- `getSubPropertySchema()`
- `isElementInSubObject()`
- `isElementInSubArray()`
- `isPropertyString()`
- `isElementInEnum()`

#### Things to note
During the refactoring process of the method, I noticed that it worked incorrectly (or at least that is my understanding). When the method is looking up the schema for an element, it incorrectly returns the schema of the `additionalInformation` element in some cases.  I do not see the point of this, and my assumption is that this is not intentional. I am however afraid of fixing this "bug", since other functionality has been built on top of this method, and is presumably "working correctly".


### The typedDataToDocumentContentWithWebform method

The `typedDataToDocumentContentWithWebform` method has been refactored and extracted from the `AtvSchema` class into its own `TypedDataToDocumentContentWithWebform` class. Previously, SonarCloud complained about the following issues regarding the method:

- The method was too long, around 520 lines.
- The method had a cognitive complexity of 350.
- The method had too much duplicate code.
- The method had if statements that could be merged.
- The method had lines that were too long.

The new `TypedDataToDocumentContentWithWebform` class and its methods fix most of these issues (the main method is now around 200 lines and has a cognitive complexity of 64). The new methods that have been implemented in the class are the following:

- `getTypedDataToDocumentContentWithWebform()` (the new "main method" of the class)
- `handlePropertyItems()`
- `handleFullItemValueCallback()`
- `modifyCallback()`
- `modifyPropertyName()`
- `isAddressField()`
- `isBankAccountField()`
- `isBudgetField()`
- `isAttachmentField()`
- `isRegularField()`
- `valueIsEmpty()`
- `getWebformElements()`
- `getValueArray()`
- `extractLabel()`
- `getFieldValuesFromPropertyItem()`
- `buildStructureArrayWithPropertyStructureCallback()`
- `extractMetadataFromWebform()`
- `writeJsonFile()` (used for testing purposes)

In addition to this, the following methods have also been moved to the new `TypedDataToDocumentContentWithWebform` class from the `AtvSchema` class, since they are purely "helper methods" for the process of converting Webform data to document content:

- `getConditionStatus()`
- `isFieldHidden()`
- `getFieldValuesFromFullItemCallback()`

#### Things to note
As with the getPropertySchema method, this method also has functionality in it that in my opinion is plainly wrong and should not be there, but I am afraid of fixing it since it might break something else. One problematic instance can be found [in the cold code ](https://github.dev/City-of-Helsinki/hel-fi-drupal-grants/blob/develop/public/modules/custom/grants_metadata/src/AtvSchema.php#L395) on the lines 577-613: In some cases variables such as $page, $section and $element are being set on iteration X of the main foreach loop, and then used in the switch statement on the line 919 on a different iteration of the main foreach loop. To me it does not seem like this is intentional, but again, since other functionality has been built on top of this method, the refactored code works in the same manner.

## How to install

Make sure your instance is up and running on the correct branch.
  * `git checkout feature/AU-1745-ATVSchema-refactor-typedDataToDocumentContentWithWebform`
  * `make fresh`
  * `make drush-cr`

## How to test
I have personally mainly been relying on the `AtvSchemaTest.php` class when testing the refactored code. The `code coverage` of both the refactored methods is good, but the `test coverage` could still be improved upon. For instance, the metadata of the elements in the `$documentStructure` array produced by the `getTypedDataToDocumentContentWithWebform` method can be filled with gibberish, and the tests will still pass. To combat this issue, I have used the `writeJsonFile()` method to produce .json files of the outputted  `$documentStructure` array, both in the `development` and the `feature/AU-1745-ATVSchema-refactor-typedDataToDocumentContentWithWebform` branch. These outputs can then be compared in a diff tool such as [jsondiff](https://www.jsondiff.com/). As of now, all the tests in `AtvSchemaTest.php `pass, and the produced .json files are identical when compared against each other. Other than that, the best way to test the refactored code is to test all the forms with different role combinations. Check that the inputted data is returned correctly from ATV after a submission has been saved as a draft.

[AU-1745]: https://helsinkisolutionoffice.atlassian.net/browse/AU-1745?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[AU-1741]: https://helsinkisolutionoffice.atlassian.net/browse/AU-1741?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
